src/BugReporting.jl

```diff
@@ -16,36 +16,90 @@ function check_rr_available()
     end
 end
 
-function do_pack(rr, trace_directory)
-    @info "Preparing trace directory for upload (if your trace is large this may take a few minutes)"
-    run(`$rr pack $trace_directory`)
+function normalize_inner_trace(trace_directory)
+    # What _RR_TRACE_DIR calls a "trace directory" is not what `rr pack` calls a "trace directory"
+    # This function allows us to normalize to the inner "latest-trace" directory, if necessary.
+    latest_symlink = joinpath(trace_directory, "latest-trace")
+    if !isfile(joinpath(trace_directory, "version")) && islink(latest_symlink)
+        return realpath(latest_symlink)
+    end
+    return trace_directory
+end
+
+function rr_pack(trace_directory)
+    check_rr_available()
+    
+    trace_directory = normalize_inner_trace(trace_directory)
+    rr() do rr_path
+        run(`$rr_path pack $(trace_directory)`)
+    end
 end
 
-function rr_record(rr)
+function is_packed(trace_directory)
+    # What _RR_TRACE_DIR calls a "trace directory" is not what `rr pack` calls a "trace directory"
+    trace_directory = normalize_inner_trace(trace_directory)
+    return !isempty(filter(f -> startswith(f, "mmap_pack"), readdir(trace_directory)))
+end
+
+function rr_record(args...; trace_dir=nothing)
+    check_rr_available()
+
     record_flags = split(get(ENV, "JULIA_RR_RECORD_ARGS", ""), ' ', keepempty=false)
-    `$rr record $(record_flags)`
+    rr() do rr_path
+        new_env = copy(ENV)
+        if trace_dir !== nothing
+            new_env["_RR_TRACE_DIR"] = trace_dir
+        end
+        # Intersperse all given arguments with spaces, then splat:
+        rr_cmd = `$(rr_path) record $(record_flags)`
+        for arg in args
+            rr_cmd = `$(rr_cmd) $(arg)`
+        end
+        run(setenv(rr_cmd, new_env))
+    end
+end
+
+function download_rr_trace(trace_url; verbose=true)
+    Pkg.PlatformEngines.probe_platform_engines!()
+    artifact_hash = Pkg.create_artifact() do dir
```

Comment:
This makes me think we should put the artifact_hash into the URL, though of course that would require some re-engineering.

Reply:
Yes, I was thinking the same thing. I think we should re-engineer to give `BugReporter` a token that allows them to upload to a directory, then we can upload not just a trace, but also a `Manifest.toml`, and potentially other information as well.  For non-x86_64-linux-gnu platforms, we of course won't have `rr` traces, but we can sketch out a format that makes sense and is somewhat consistent across platforms.  I'll open a new issue.